### PR TITLE
New feature: autoload: package-lint-package, lint multiple files

### PR DESCRIPTION
**Pull request by Boruch-Baum**
+ lints all files matching a name-space in a specific directory

+ lints all files in a directory tree

+ configurable subdirectory exclusion list, by regexp

  + package-lint-package-exclude-subdirectories

### CODE REVIEW

**Comment by conao3** [comment]
Nice work!

File: package-lint.el [hunk 1]
```diff
@@ -7,7 +7,7 @@
 ;; URL: https://github.com/purcell/package-lint
 ;; Keywords: lisp
 ;; Version: 0
-;; Package-Requires: ((cl-lib "0.5") (emacs "24.1") (let-alist "1.0.6"))
+;; Package-Requires: ((emacs "25.1") (let-alist "1.0.6"))
```
Comment (conao3) [inline package-lint.el:10]: Why are you change minimum required Emacs version?
If you really want to this change, please create another PR.

Comment (Boruch-Baum) [reply package-lint.el:10]: I performed package-lint on my modifications to package-lint and it gave me warnings that several of the functions I was using were from later Emacs versions. My memory is that function `directory-files-recursively' was what required 25.1. I thought to keep in the same PR because otherwise package-lint would fail itself, which would be embarrassing and ironic!

**Comment by akirak** [comment]
@Boruch-Baum Because there can be a multi-package repository, i.e. a repository that contains source files for multiple packages in the same directory, you'll need a MELPA-style recipe to properly determine the files of a package. 

Both riscy's [melpazoid](https://github.com/riscy/melpazoid) and conao3's [keg.el](https://github.com/conao3/keg.el) do this, so there are already multiple solutions that run package-lint on an entire package, even though all of them require some additional configuration.

[package-build](https://github.com/melpa/melpa/blob/master/package-build/package-build.el) provides a function to retrieve a list of files from a directory and a recipe, which you could use to implement the proper version of `package-lint-package`. However, this means you will have to add `package-build` to the dependencies of `package-lint`.

**Comment by Boruch-Baum** [comment]
On 2021-09-01 07:35, Akira Komamura wrote:
> ***@***.*** Because there can be a multi-package repository, i.e. a
> repository that contains source files for multiple packages in the same
> directory, you'll need a MELPA-style recipe to properly determine the
> files of a package.

Fundamentally, I'm not sure that's so relevant. The people most likely
to lint a repository are going to be the developers themselves, who I'm
imagining will have mostly be interested in linting one file at a time
or everything. They may not want to be part of the MELPA eco-system and
so that type of recipe might not exist.

Even for a MELPA curator, I'm not sure that your claim is true. With
this PR, if you select a file you only lint that file's name-space, not
the entire directory/repository. That should cover just the individual
package's elements within the repository, no?

If it does turn out to be the case that occasionally there is a
conflict, how common would that be? What percentage of cases does this
PR need to fail for it not to be an improvement? In this regard, I'd
distinguish between false-positives and false-negatives: A
false-positive I would consider a small pain, but easy to ignore;
false-negatives I would consider a nuisances. However, if it does need
to be 'hypothetically perfect' instead of 'great in practice', then I
guess the PR will be dropped!

Right now without a real-life example to test, my talk is hypothetical
Do you have a real-life failure example in mind that you can offer?

> Both riscy's [2]melpazoid and conao3's [3]keg.el do this, so there are
> already multiple solutions that run package-lint on an entire package,

I didn't know that! The only reason I made the effort is because on
another thread someone indicated that the feature did not exist ...

I'm not familiar with either melpazoid or keg ... Taking a first look,
it doesn't seem fair to compare package-lint to either of them. They may
be great, but ... melpazoid isn't even a MELPA package ?! And keg looks
like some kind of wishful parallel universe, with a very small adoption
via MELPA ...

> even though all of them require some additional configuration.

Ah. As so often the case, the devil is in the details ...

>  [4]package-build provides a function to retrieve a list of files from
>  a directory and a recipe, which you could use to implement the proper
>  version of package-lint-package. However, this means you will have to
>  add package-build to the dependencies of package-lint.

I'm not the package owner, but my sensibility would be to "go lean" if
at all possible and avoid that. If @purcell doesn't like the PR, that's
fine; he didn't ask me to do the work, so the effort is on me. Maybe it
can be the basis for someone else to do a better job.

To sum up: Can you point me to a problematic example?

--
hkp://keys.gnupg.net
CA45 09B5 5351 7C11 A9D1  7286 0036 9E45 1595 8BC0

**Comment by akirak** [comment]
See also #111 for the previous discussions on support for multiple files.

**Comment by Boruch-Baum** [comment]
On 2021-09-01 19:54, Akira Komamura wrote:
>    See also [1]#111 for the previous discussions on support for multiple
>    files.

Thanks. I'm being side-tracked just now by something urgent, so I'm
going to be busy with that for a bit, but thanks for that and your
previous mail about key that I also haven't been able to study.

--
hkp://keys.gnupg.net
CA45 09B5 5351 7C11 A9D1  7286 0036 9E45 1595 8BC0
